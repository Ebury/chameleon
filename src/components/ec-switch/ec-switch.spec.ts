import { type ComponentMountingOptions, mount } from '@vue/test-utils';
import { vi } from 'vitest';
import { defineComponent, ref } from 'vue';

import EcSwitch from './ec-switch.vue';
import type { SwitchProps } from './types';

describe('EcSwitch', () => {
  function mountSwitch(props?: SwitchProps, mountOpts?: ComponentMountingOptions<typeof EcSwitch>) {
    return mount(EcSwitch, {
      props,
      ...mountOpts,
    });
  }

  describe(':props', () => {
    it(':label - should render the switch with a label', () => {
      const wrapper = mountSwitch({
        label: 'Test label prop',
      });

      expect(wrapper.findByDataTest('ec-switch__label').exists()).toBe(true);
      expect(wrapper.element).toMatchSnapshot();
    });

    it(':error-message - should render the switch with an error message', () => {
      const wrapper = mountSwitch({
        errorMessage: 'Test error message prop',
      });

      expect(wrapper.findByDataTest('ec-switch__error-text').exists()).toBe(true);
      expect(wrapper.element).toMatchSnapshot();
    });

    it(':disabled - should render the switch with a disabled attribute and not checked', () => {
      const wrapper = mountSwitch({
        disabled: true,
        modelValue: false,
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it(':disabled - should render the switch with a disabled attribute and checked', () => {
      const wrapper = mountSwitch({
        disabled: true,
        modelValue: true,
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    describe('#slots', () => {
      it('#errorMessage - should render the error message slot if passed', () => {
        const wrapper = mountSwitch(
          {},
          {
            slots: {
              'error-message': 'Test error message slot',
            },
          },
        );

        expect(wrapper.findByDataTest('ec-switch__error-text').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });

      it('#error-message - should render error-message slot instead of error-message prop when both are passed', () => {
        const wrapper = mountSwitch(
          {
            errorMessage: 'Error message coming from props',
          },
          {
            slots: {
              'error-message': 'Error message coming from slot',
            },
          },
        );

        expect(wrapper.findByDataTest('ec-switch__error-text').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });

      it('#label - should render the label slot if passed', () => {
        const wrapper = mountSwitch(
          {},
          {
            slots: {
              label: 'Test label message slot',
            },
          },
        );

        expect(wrapper.findByDataTest('ec-switch__label').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });

      it('#label - should render label slot instead of label prop when both are passed', () => {
        const wrapper = mountSwitch(
          {
            label: 'Label message coming from props',
          },
          {
            slots: {
              label: 'Label message coming from slot',
            },
          },
        );

        expect(wrapper.findByDataTest('ec-switch__label').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('@events', () => {
      it('@update:modelValue - should be emitted when input is clicked', () => {
        const wrapper = mountSwitch();

        wrapper.findByDataTest('ec-switch__input').setValue(true);
        expect(wrapper.emitted('update:modelValue')?.length).toBe(1);
      });

      it('@change - should be emitted when input is clicked', () => {
        const changeSpy = vi.fn();
        const wrapper = mountSwitch({}, {
          attrs: {
            onChange: changeSpy,
          },
        });

        wrapper.findByDataTest('ec-switch__input').setValue(true);
        expect(changeSpy).toHaveBeenCalledTimes(1);
      });
    });

    describe('attrs', () => {
      it('should pass all non-prop attributes to the hidden switch input', () => {
        const wrapper = mountSwitch(
          {},
          {
            attrs: {
              'aria-label': 'Random label',
            },
          },
        );
        expect(wrapper.element).toMatchSnapshot();
      });

      it('should render with custom attributes', () => {
        const wrapper = mountSwitch({}, {
          attrs: {
            'data-test': 'my-data-test',
            class: 'my-class',
            id: 'test-id',
            style: 'top: 0px',
            role: 'switch',
          },
        });
        expect(wrapper.element).toMatchSnapshot();
      });

      it('should pass listeners to switch input', async () => {
        const clickSpy = vi.fn();
        const wrapper = mountSwitch({}, {
          attrs: { onClick: clickSpy },
        });

        await wrapper.findByDataTest('ec-switch__input').trigger('click');
        expect(clickSpy).toHaveBeenCalledTimes(1);
      });
    });

    describe('v-model', () => {
      it('should render the switch and toggle v-model value when input is clicked', async () => {
        const Component = defineComponent({
          components: { EcSwitch },
          setup() {
            const checked = ref(true);
            return { checked };
          },
          template: '<ec-switch v-model="checked"></ec-switch>',
        });

        const wrapper = mount(Component);

        expect(wrapper.findByDataTest('ec-switch').exists()).toBe(true);
        expect(wrapper.vm.checked).toBe(true);
        await wrapper.findByDataTest('ec-switch__input').setValue(false);
        expect(wrapper.vm.checked).toBe(false);
      });

      it('should not change the value of v-model if disabled is enabled and user clicks the switch', async () => {
        const Component = defineComponent({
          components: { EcSwitch },
          setup() {
            const checked = ref(true);
            return { checked };
          },
          template: '<ec-switch v-model="checked" disabled></ec-switch>',
        });

        const wrapper = mount(Component);

        expect(wrapper.vm.checked).toBe(true);
        await wrapper.findByDataTest('ec-switch__input').setValue(false);
        expect(wrapper.vm.checked).toBe(true);
      });

      it('should be clickable', async () => {
        const element = document.createElement('div');
        document.body.appendChild(element);

        const Component = defineComponent({
          components: { EcSwitch },
          data() {
            return { model: false };
          },
          template: '<ec-switch v-model="model" />',
        });

        const wrapper = mount(Component, {
          attachTo: element,
        });

        await wrapper.findByDataTest('ec-switch__wrapper').trigger('click');
        expect(wrapper.vm.model).toBe(true);

        await wrapper.findByDataTest('ec-switch__wrapper').trigger('click');
        expect(wrapper.vm.model).toBe(false);

        wrapper.unmount();
      });
    });
  });
});
