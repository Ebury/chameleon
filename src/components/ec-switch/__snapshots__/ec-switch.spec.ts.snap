// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EcSwitch > :props > #slots > #error-message - should render error-message slot instead of error-message prop when both are passed 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    aria-describedby="ec-switch-error-1"
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper ec-switch__wrapper--error"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <div
    class="ec-switch__error-text"
    data-test="ec-switch__error-text"
    id="ec-switch-error-1"
  >
    
    Error message coming from slot
    
  </div>
</div>
`;

exports[`EcSwitch > :props > #slots > #errorMessage - should render the error message slot if passed 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    aria-describedby="ec-switch-error-1"
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper ec-switch__wrapper--error"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <div
    class="ec-switch__error-text"
    data-test="ec-switch__error-text"
    id="ec-switch-error-1"
  >
    
    Test error message slot
    
  </div>
</div>
`;

exports[`EcSwitch > :props > #slots > #label - should render label slot instead of label prop when both are passed 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title="Label message coming from props"
    >
      
      Label message coming from slot
      
    </label>
  </div>
  <!--v-if-->
</div>
`;

exports[`EcSwitch > :props > #slots > #label - should render the label slot if passed 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      Test label message slot
      
    </label>
  </div>
  <!--v-if-->
</div>
`;

exports[`EcSwitch > :props > :disabled - should render the switch with a disabled attribute and checked 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    class="ec-switch__input"
    data-test="ec-switch__input"
    disabled=""
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper ec-switch__wrapper--checked ec-switch__wrapper--disabled ec-switch__wrapper--checked-and-disabled"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <!--v-if-->
</div>
`;

exports[`EcSwitch > :props > :disabled - should render the switch with a disabled attribute and not checked 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    class="ec-switch__input"
    data-test="ec-switch__input"
    disabled=""
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper ec-switch__wrapper--disabled"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <!--v-if-->
</div>
`;

exports[`EcSwitch > :props > :error-message - should render the switch with an error message 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    aria-describedby="ec-switch-error-1"
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper ec-switch__wrapper--error"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <div
    class="ec-switch__error-text"
    data-test="ec-switch__error-text"
    id="ec-switch-error-1"
  >
    
    Test error message prop
    
  </div>
</div>
`;

exports[`EcSwitch > :props > :label - should render the switch with a label 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title="Test label prop"
    >
      
      Test label prop
      
    </label>
  </div>
  <!--v-if-->
</div>
`;

exports[`EcSwitch > :props > attrs > should pass all non-prop attributes to the hidden switch input 1`] = `
<div
  class="ec-switch"
  data-test="ec-switch"
>
  <input
    aria-label="Random label"
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <!--v-if-->
</div>
`;

exports[`EcSwitch > :props > attrs > should render with custom attributes 1`] = `
<div
  class="ec-switch my-class"
  data-test="my-data-test ec-switch"
  style="top: 0px;"
>
  <input
    class="ec-switch__input"
    data-test="ec-switch__input"
    id="ec-switch-1"
    role="switch"
    type="checkbox"
  />
  <div
    class="ec-switch__label-switch-wrapper"
  >
    <span
      class="ec-switch__wrapper"
      data-test="ec-switch__wrapper"
    />
    <label
      class="ec-switch__label"
      data-test="ec-switch__label"
      for="ec-switch-1"
      title=""
    >
      
      
      
    </label>
  </div>
  <!--v-if-->
</div>
`;
